File: src/main/java/dev/kikugie/elytratrims/common/recipe/ETPatternRecipe.java
package dev.kikugie.elytratrims.common.recipe;

import dev.kikugie.elytratrims.common.ETServer;
import dev.kikugie.elytratrims.common.access.FeatureAccess;
import net.minecraft.block.entity.BannerBlockEntity;
import net.minecraft.inventory.Inventory;
import net.minecraft.item.BannerItem;
import net.minecraft.item.ItemStack;
import net.minecraft.recipe.RecipeSerializer;
import net.minecraft.recipe.SpecialRecipeSerializer;
import net.minecraft.recipe.book.CraftingRecipeCategory;
import net.minecraft.util.Identifier;

public class ETPatternRecipe extends DelegatedRecipe {
    public static RecipeSerializer<ETPatternRecipe> SERIALIZER = new SpecialRecipeSerializer<>(ETPatternRecipe::new);
    public ETPatternRecipe(Identifier id, CraftingRecipeCategory category) {
        super(id, category);
    }

    @Override
    boolean matches(Inventory inventory) {
        int elytra = 0;
        int banner = 0;
        for (int slot = 0; slot < inventory.size(); slot++) {
            ItemStack stack = inventory.getStack(slot);
            if (ETServer.isProbablyElytra(stack.getItem())) elytra++;
            else if (stack.getItem() instanceof BannerItem) {
                if (BannerBlockEntity.getPatternCount(stack) == 0) return false;
                banner++;
            } else if (!stack.isEmpty()) return false;
            if (elytra > 1 || banner > 1) return false;
        }
        return elytra == 1 && banner == 1;
    }

    @Override
    ItemStack craft(Inventory inventory) {
        ItemStack elytra = ItemStack.EMPTY;
        ItemStack banner = ItemStack.EMPTY;
        for (int slot = 0; slot < inventory.size(); slot++) {
            ItemStack stack = inventory.getStack(slot);
            if (stack.isEmpty()) continue;
            if (ETServer.isProbablyElytra(stack.getItem())) elytra = stack.copy();
            else if (stack.getItem() instanceof BannerItem) banner = stack;
        }
        FeatureAccess.setPatterns(elytra, banner);
        return elytra;
    }

    @Override
    public boolean fits(int width, int height) {
        return width * height >= 2;
    }

    @Override
    public RecipeSerializer<?> getSerializer() {
        return null;
    }
}

File: src/main/kotlin/dev/kikugie/elytratrims/common/recipe/ETPatternRecipe.kt
package dev.kikugie.elytratrims.common.recipe

import dev.kikugie.elytratrims.api.ElytraTrimsAPI
import dev.kikugie.elytratrims.common.access.FeatureAccess.getBaseColor
import dev.kikugie.elytratrims.common.access.FeatureAccess.getPatterns
import dev.kikugie.elytratrims.common.access.FeatureAccess.setColor
import dev.kikugie.elytratrims.common.access.FeatureAccess.setPatterns
import net.minecraft.item.BannerItem
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.recipe.RecipeSerializer
import net.minecraft.util.Identifier

class ETPatternRecipe(id: Identifier) : DelegatedRecipe(id, SAMPLE)  {
    override fun matches(input: Stacks): Boolean {
        var elytra = 0
        var banner = 0
        input.forEach {
            if (ElytraTrimsAPI.isProbablyElytra(it)) elytra++
            else if (it.item is BannerItem) {
                if (it.getPatterns().isEmpty()) return false
                banner++
            } else if (!it.isEmpty) return false
            if (elytra > 1 || banner > 1) return false
        }
        return elytra == 1 && banner == 1
    }

    override fun craft(input: Stacks): ItemStack {
        val banner = input.firstItemCopy { it is BannerItem } ?: return ItemStack.EMPTY
        val elytra = input.firstItemCopy(ElytraTrimsAPI::isProbablyElytra) ?: return ItemStack.EMPTY
        elytra.setPatterns(banner)
        val color = banner.getBaseColor()
        if (color != 0) elytra.setColor(color)
        return elytra
    }

    override fun fits(width: Int, height: Int): Boolean {
        return width * height >= 2
    }

    companion object {
        val SAMPLE = ItemStack(Items.ELYTRA).apply {
            // TODO
        }
    }
}